e2e tests for multizone PDs

verult opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**: e2e tests for multizone PDs.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
partially fixing #59988

/release-note-none

/assign @msau42 @saad-ali

verult committed:
multi-zone PD e2e tests

msau42 commented on test/e2e/storage/regional_pd.go:
```diff
@@ -0,0 +1,346 @@
+		provisioner:    "kubernetes.io/gce-pd",
+		parameters: map[string]string{
+			"type":          "pd-standard",
+			"replica-zones": strings.Join(cloudZones, ","),
```
Also a test case for "auto"?

msau42 commented on test/e2e/storage/regional_pd.go:
```diff
@@ -0,0 +1,346 @@
+		claimSize:    "1.5G",
+		expectedSize: "2G",
+		pvCheck: func(volume *v1.PersistentVolume) error {
+			return checkGCEPD(volume, "pd-standard")
```
Do you also want to verify that the PD has the correct zone labels on it?  Also, any way we can validate from cloud provider that it's a RePD?

verult replied:
Will add code to check zone labels. No I can't think of an easy way to check for RePDs. `GCECloud` doesn't expose a `GetDisk()` today (maybe it should for testing purposes?).

msau42 commented on test/e2e/storage/regional_pd.go:
```diff
@@ -0,0 +1,346 @@
+			"Error waiting for nodes from the new instance group to become ready.")
+	}()
+
+	// Rescheduling of the pod should occur while instance group deletion is in progress.
```
Yeah I would wait and validate that the pod has restarted to avoid possible race conditions.

msau42 commented on test/e2e/storage/regional_pd.go:
```diff
@@ -0,0 +1,346 @@
+		"Error getting logs from pod %s in namespace %s", pod.Name, ns)
+	lineCount := len(strings.Split(strings.TrimSpace(logs), "\n"))
+	expectedLineCount := 2
+	Expect(lineCount).To(Equal(expectedLineCount),
```
Would it be too fancy for the Pod to print out the zone it's in, and validate that?

verult replied:
Is there a way for a pod to get its zone information other than to `kubectl get node` and check its label? We are already kind of doing that in the test.

msau42 replied:
I'm not sure if downward api can expose labels.

verult replied:
Had a chat with @msau42 offline - pods usually don't expose zone labels by default, so it's a bit more involved to get zone info within a container.

verult commented:
/release-note-none

verult commented:
Addressed comments. Also updated the test to use the newly proposed storageclass params. Placing a hold here until PR for updated SC params are merged.

/hold

msau42 commented on test/e2e/e2e.go:
```diff
@@ -72,7 +72,10 @@ func setupProviderConfig() error {
 		}
 
-		gceAlphaFeatureGate, err := gcecloud.NewAlphaFeatureGate([]string{gcecloud.AlphaFeatureNetworkEndpointGroup})
+		gceAlphaFeatureGate, err := gcecloud.NewAlphaFeatureGate([]string{
```
If you intend to submit this as alpha, you will also have to whitelist the alpha test project to be able to use the alpha disk api

verult replied:
Yep, I intend to submit this as beta, but will do this for the regular test project

msau42 replied:
For beta you don't need to whitelist right?

verult replied:
We do

msau42 commented on test/e2e/framework/util.go:
```diff
@@ -5163,3 +5191,17 @@ func WaitForPersistentVolumeClaimDeleted(c clientset.Interface, ns string, pvcNa
 	return fmt.Errorf("PersistentVolumeClaim %s is not removed from the system within %v", pvcName, timeout)
 }
+
+func GetClusterZones(c clientset.Interface) sets.String {
```
We already have other e2e tests that get all the zones in the cluster.  Can you use the same api they're using instead?

verult replied:
Talked to @msau42 offline. Most tests re-implement this functionality with the same way of getting zone info. I'm changing some of them to use this method instead.

msau42 commented on test/e2e/storage/regional_pd.go:
```diff
@@ -0,0 +1,438 @@
+})
+
+func testVolumeProvisioning(c clientset.Interface, ns string) {
+	cloudZones := getRandomCloudZones(c)
```
getTwoRandomZones?

msau42 commented on test/e2e/storage/regional_pd.go:
```diff
@@ -0,0 +1,438 @@
+	err = framework.WaitForStatefulSetReplicasReady(statefulSet.Name, ns, c, framework.Poll, statefulSetReadyTimeout)
+	if err != nil {
+		pod := getPod(c, ns, repdLabels)
+		Expect(podutil.IsPodReadyConditionTrue(pod.Status)).To(BeTrue(),
```
Is this for more debugging?  Could you just Expect(err).NotTo(HaveOccurred())?

verult replied:
Yeah. The error just indicates there was a timeout waiting for statefulset to become ready. In the case the test fails due to an actual bug, it's helpful to print out the pod state.

msau42 commented on test/e2e/storage/regional_pd.go:
```diff
@@ -0,0 +1,438 @@
+
+	pvc := getPVC(c, ns, repdLabels)
+	defer func() {
+		framework.Logf("deleting statefulset%q/%q", statefulSet.Namespace, statefulSet.Name)
```
Do you need to delete the service too?

msau42 commented on test/e2e/storage/regional_pd.go:
```diff
@@ -0,0 +1,438 @@
+	instanceGroup, err := cloud.GetInstanceGroup(instanceGroupName, podZone)
+	Expect(err).NotTo(HaveOccurred(),
+		"Error getting instance group %s in zone %s", instanceGroupName, podZone)
+	err = framework.DeleteManagedInstanceGroup(podZone)
```
You already have the instanceGroupName here, could you just pass that in, instead of the zone name?

verult replied:
DeleteManagedInstanceGroup() already has the instance group name (from the TestContext), but the same name is used in every zone, so it needs the zone name to identify the right instance group.

verult replied:
We do, because it's used to get the instance group through the cloud provider API, which has to happen before deleting the group.

msau42 commented on test/e2e/storage/regional_pd.go:
```diff
@@ -0,0 +1,441 @@
+}
+
+// Waits for at least 1 replica of a StatefulSet to become not ready or until timeout occurs, whichever comes first.
+func waitForStatefulSetReplicasNotReady(statefulSetName, ns string, c clientset.Interface) error {
```
Have you checked if there are already existing statefulset utility functions?  I leveraged a few for a local PV test.

verult commented:
Addressed all comments except changing these tests to use existing statefulset functions. Also fixed a bug where the zone super set was calculated incorrectly in `verifyZonesInPV()`

verult commented:
Removed the `[Feature]` tag to allow these test be run in normal CI environment

msau42 commented on test/e2e/storage/regional_pd.go:
```diff
@@ -0,0 +1,445 @@
+
+	pvc := getPVC(c, ns, regionalPDLabels)
+	defer func() {
+		framework.Logf("deleting statefulset%q/%q", statefulSet.Namespace, statefulSet.Name)
```
Should you defer the statefulset and service deletion right after creating?

msau42 reviewed (commented):
In general this lgtm.  If you're going to wait to submit until beta, then you can remove those alpha changes.

verult commented:
Current plan is to submit this PR after the StorageClass PR (https://github.com/kubernetes/kubernetes/pull/60165) but before the alpha gate change, so I'll leave the alpha gate here as is and remove it in the beta PR.

msau42 commented:
/lgtm
/approve no-issue

verult commented:
/retest

saad-ali commented:
/approve

verult commented:
#60165 is merged

/hold cancel
